pkgs/top-level/named-hierarchy-files.nix

```
@@ -0,0 +1,108 @@
+/*
+Internal Nixpkgs file, it may be removed or changed at any point without warning
+
+This file contains the mapping from attribute name to the file in pkgs/by-name that contains a definition of it.
+No validity checks are performed, instead this file is optimized for performance.
+TODO: Link to manual and validity checks
+
+Example: If `pkgs/by-name/he/hello` exists, the result will have a `hello = pkgs/by-name/he/hello/package.nix` entry.
+
+Type: AttrsOf Path
+*/
+
+# Because the result might be fairly large, we only want compute it once.
+# In order to ensure this, we make use of Nix's `import` cache.
+# This cache memoizes the resulting value of the imported file, notably including any `let in` bindings at the top of the file.
+# This is why everything here is computed in a `let in` binding at the top.
+let
+
+  # We don't _need_ `lib` here, but importing it wouldn't be detrimental to performance either because of Nix's `import` cache
+  inherit (builtins) attrNames elemAt length mapAttrs readDir;
+
+  # Warning: Do not make this configurable, otherwise the `import` evaluation cache won't work!
+  baseDirectory = ../by-name;
+
+  entries = readDir baseDirectory;
+  shards = attrNames entries;
+
+  # Return the name mappings in a single shard
+  # Type: String -> AttrsOf Path
+  namesForShard = shard:
+    if entries.${shard} != "directory" then
+      # Ignore non-directories. Technically only Readme.md is allowed as a file in the base directory, so we could alternatively:
+      # - Assume that Readme.md is the only file and change the condition to `shard == "Readme.md"` for a minor performance improvement.
+      #   This would however cause very poor error messages if there's other files.
+      # - Ensure that Readme.md is the only file, throwing a better error message if that's not the case.
+      #   However for good code architecture, all error messages relating to `pkgs/by-name` should be centralized in (TODO: Link to validity checks)
+      # Additionally in either of those alternatives, we would have to duplicate the hardcoding of "Readme.md"
+      { }
+    else
+      mapAttrs
+        (name: _: baseDirectory + "/${shard}/${name}/package.nix")
+        (readDir (baseDirectory + "/${shard}"));
+
+  # Merge the name mappings of all shards between the indices start (inclusive) and end (exclusive)
+  # This ignores any potential duplicates for efficiency
+  # Invariant: start < end
+  # Type: Int -> Int -> AttrsOf Path
+  binaryMerge = start: end:
```

```suggestion
  # Type: Int -> Int -> AttrsOf Path
  binaryMerge = start: end:
    # Invariant: start < end
```

Names aren't part of the public interface of a function. (attribute names would be)

---

Not sure what you mean to say with that, the suggestion to move the invariant into the function sounds okay though

---

I guess I'm extrapolating from
 - name changes are not observable by callers
 - names used in a function implementation aren't always optimal, either for brevity or to avoid shadowing something in their scope

Combining those, you may conclude that argument names should not be considered as part of the public interface, although it is somewhat subjective.

I suppose you could use an `assert`, or a commented out one so we don't have to think too much about performance

```suggestion
  # Type: Int -> Int -> AttrsOf Path
  binaryMerge = start: end:
    # assert start < end;
```

---

Ohh I get it now, the problem is that the invariant uses variables available only inside the function. Makes sense, I'll apply this suggestion